etl/steps/data/garden/homicide/2023-01-04/unodc.py

```diff
@@ -76,13 +76,55 @@ def harmonize_countries(df: pd.DataFrame) -> pd.DataFrame:
 
 
 def clean_data(df: pd.DataFrame) -> pd.DataFrame:
+    # Making the units more obvious
+    df["unit_of_measurement"] = df["unit_of_measurement"].replace(
+        ["Counts", "Rate per 100,000 population"], ["Homicide count", "Homicides per 100,000 population"]
+    )
+    # Splitting the data into that which has the totals and that which is disaggregated by mechanism
+    df_mech = df[df["dimension"] == "by mechanisms"]
 
-    df = df[["country", "year", "unit_of_measurement", "value"]]
+    df_mech = create_mechanism_df(df_mech)
 
-    df = (
-        df.pivot(index=["country", "year"], columns="unit_of_measurement", values="value")
-        .reset_index()
-        .rename_axis(None, axis=1)
-    )
+    df_tot = df[df["dimension"] == "Total"]
+
+    df_tot = create_total_df(df_tot)
+
+    df = pd.merge(df_mech, df_tot, how="outer", on=["country", "year"])
 
     return df
+
+
+def create_total_df(df_tot: pd.DataFrame) -> pd.DataFrame:
+    """Create the total homicides dataframe where we will have total homicides/homicide rate
+    disaggregated by age and sex
+    """
+    df_tot = df_tot.drop(columns=["region", "subregion", "indicator", "dimension", "category", "source"])
+    # There are some duplicates when sex is unknown so let's remove those rows
+    df_tot = df_tot[df_tot["sex"] != "Unknown"]
+    # Make it more obvious what total age and total sex means
+    df_tot["age"] = df_tot["age"].replace("Total", "All ages")
+    df_tot["sex"] = df_tot["sex"].replace("Total", "Both sexes")
+
+    df_tot = df_tot.pivot(index=["country", "year"], values=["value"], columns=["sex", "age", "unit_of_measurement"])
+    # Make the columns nice
+    df_tot.columns = df_tot.columns.droplevel(0)
+    df_tot.columns = df_tot.columns.map("_".join)
+    df_tot = df_tot.reset_index()
+    return df_tot
+
+
+def create_mechanism_df(df_mech: pd.DataFrame) -> pd.DataFrame:
+    """Create the homicides by mechanism dataframe where we will have  homicides/homicide rate
+    disaggregated by mechanism (e.g. weapon)
+    """
+    df_mech = df_mech.drop(columns=["region", "subregion", "indicator", "dimension", "source", "sex", "age"])
+    df_mech["category"] = df_mech["category"].replace(
+        ["Firearms or explosives - firearms", "Another weapon - sharp object"], ["Firearms", "Sharp object"]
+    )
+    # Make the table wider so we have a column for each mechanism
+    df_mech = df_mech.pivot(index=["country", "year"], values=["value"], columns=["category", "unit_of_measurement"])
+    # Make the columns nice
+    df_mech.columns = df_mech.columns.droplevel(0)
+    df_mech.columns = df_mech.columns.map("_".join)
+    df_mech = df_mech.reset_index()
```

Comment:
Seems that this logic is written twice (for `create_total_df` and `create_mechanism_df`), perhaps you can have an internal function, sth like:

```python
def _some_name(df, drop_columns, pivot_index, pivot_values, pivot_columns):
        df = df.drop(columns=drop_columns)
	df = df.pivot(
		index=pivot_index,
		values=pivot_values,
		columns=pivot_columns,
	)
	# Make the columns nice
	df.columns = df.columns.droplevel(0)
	df.columns = df.columns.map("_".join)
	df = df.reset_index()
        return df
```

So overall, the function could look like:

```python
def create_mechanism_df(df_mech: pd.DataFrame) -> pd.DataFrame:
    """Create the homicides by mechanism dataframe where we will have  homicides/homicide rate
    disaggregated by mechanism (e.g. weapon)
    """
    df_mech = df_mech.drop(columns=["region", "subregion", "indicator", "dimension", "source", "sex", "age"])
    df_mech["category"] = df_mech["category"].map({
        "Firearms or explosives - firearms": "Another weapon - sharp object",
        "Firearms": "Sharp object"
    }, na_action="ignore")
    df_mech = _create_df(
    	df_mech,
    	["region", "subregion", "indicator", "dimension", "source", "sex", "age"]
        ["country", "year"],
        ["value"],
        ["category", "unit_of_measurement"]
    )
    return df_mech
```

Reply:
Oo thank you!